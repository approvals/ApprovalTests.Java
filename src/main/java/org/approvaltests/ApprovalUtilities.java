package org.approvaltests;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class ApprovalUtilities
{
  public ByteArrayOutputStream writeSystemOutToStringBuffer()
  {
    ByteArrayOutputStream text = new ByteArrayOutputStream();
    System.setOut(new PrintStream(text));
    return text;
  }
}
