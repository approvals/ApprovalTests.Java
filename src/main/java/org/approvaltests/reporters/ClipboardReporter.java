package org.approvaltests.reporters;

import java.awt.GraphicsEnvironment;
import java.io.File;

import com.spun.util.SystemUtils;
import com.spun.util.WindowUtils;

/**
 * A reporter which creates the command to accept the received
 * file as the approve file and copies it to the clipboard:<br>
 * <code>move received.txt approved.txt</code>
 */
public class ClipboardReporter implements EnvironmentAwareReporter
{
  @Override
  public void report(String received, String approved) throws Exception
  {
    WindowUtils.copyToClipBoard(getCommandLine(received, approved), false);
  }
  public static String getCommandLine(String received, String approved)
  {
    File r = new File(received);
    File a = new File(approved);
    String commandLine = ClipboardReporter.getAcceptApprovalText(r.getAbsolutePath(), a.getAbsolutePath());
    return commandLine;
  }
  /**
   * The clipboard will not be available in a headless environment.
   */
  @Override
  public boolean isWorkingInThisEnvironment(String forFile)
  {
    return !GraphicsEnvironment.isHeadless();
  }
  public static String getAcceptApprovalText(String received, String approved)
  {
    if (SystemUtils.isWindowsEnviroment())
    {
      return String.format("move /Y \"%s\"  \"%s\"", received, approved);
    }
    else
    {
      return String.format("mv %s %s", received, approved);
    }
  }
}
