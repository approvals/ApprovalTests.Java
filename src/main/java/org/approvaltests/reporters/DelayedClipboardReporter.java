package org.approvaltests.reporters;

import java.awt.GraphicsEnvironment;

import com.spun.util.StringUtils;
import com.spun.util.WindowUtils;

/**
 * A reporter which creates the command to accept the received
 * file as the approve file and copies it to the clipboard:<br>
 * <code>move received.txt approved.txt</code>
 */
public class DelayedClipboardReporter implements EnvironmentAwareReporter
{
  private static StringBuffer text = new StringBuffer();
  @Override
  public void report(String received, String approved) throws Exception
  {
    String commandLine = ClipboardReporter.getCommandLine(received, approved);
    text.append(commandLine + StringUtils.NEW_LINE);
    WindowUtils.copyToClipBoard(text.toString(), false);
  }
  /**
   * The clipboard will not be available in a headless environment.
   */
  @Override
  public boolean isWorkingInThisEnvironment(String forFile)
  {
    return !GraphicsEnvironment.isHeadless();
  }
}
