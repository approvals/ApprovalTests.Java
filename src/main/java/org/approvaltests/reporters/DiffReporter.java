package org.approvaltests.reporters;

import org.approvaltests.reporters.macosx.MacDiffReporter;
import org.approvaltests.reporters.windows.WindowsDiffReporter;

public class DiffReporter extends FirstWorkingReporter
{
  public static final DiffReporter INSTANCE = new DiffReporter();
  public DiffReporter()
  {
    super(WindowsDiffReporter.INSTANCE, MacDiffReporter.INSTANCE, JunitReporter.INSTANCE, QuietReporter.INSTANCE);
  }
}
