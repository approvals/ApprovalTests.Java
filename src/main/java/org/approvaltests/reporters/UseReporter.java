package org.approvaltests.reporters;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import org.approvaltests.core.ApprovalFailureReporter;

@Retention(RetentionPolicy.RUNTIME)
public @interface UseReporter {
  Class<? extends ApprovalFailureReporter>[] value();
}
