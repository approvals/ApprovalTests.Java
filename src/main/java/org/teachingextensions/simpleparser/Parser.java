package org.teachingextensions.simpleparser;

import java.lang.reflect.Field;
import java.util.List;

import org.lambda.functions.Function1;
import org.lambda.query.Query;

import com.spun.util.ObjectUtils;
import com.spun.util.io.FileUtils;

public class Parser
{
  public static String parse(String text, Object data)
  {
    return parse(text, "{", "}", data);
  }
  public static String parse(String text, String startCharacter, String endCharacter, Object data)
  {
    StringBuffer sb = new StringBuffer();
    String temp = text;
    int start = temp.indexOf(startCharacter);
    while (start > 0)
    {
      int end = temp.indexOf(endCharacter, start);
      sb.append(temp.substring(0, start));
      String key = temp.substring(start + startCharacter.length(), end);
      sb.append(getValue(key, data));
      temp = temp.substring(end + endCharacter.length());
      start = temp.indexOf(startCharacter);
    }
    sb.append(temp);
    return sb.toString();
  }
  private static String getValue(final String piece, Object data)
  {
    try
    {
      Field[] declaredFields = data.getClass().getDeclaredFields();
      List<Field> f = Query.where(declaredFields, new Function1<Field, Boolean>()
      {
        @Override
        public Boolean call(Field i)
        {
          return i.getName().equals(piece);
        }
      });
      f.get(0).setAccessible(true);
      return "" + f.get(0).get(data);
    }
    catch (Exception e)
    {
      return piece;
    }
  }
  public static String parseRtfFile(String fileName, Object data)
  {
    try
    {
      String text = FileUtils.readFromClassPath(data.getClass(), fileName);
      return parse(text, "\\{", "\\}", data);
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
}
