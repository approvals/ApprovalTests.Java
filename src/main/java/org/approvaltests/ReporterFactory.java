package org.approvaltests;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.FileLauncherReporter;
import org.approvaltests.reporters.ImageReporter;
import org.approvaltests.reporters.MultiReporter;
import org.approvaltests.reporters.QuietReporter;
import org.approvaltests.reporters.UseReporter;

import com.spun.util.ClassUtils;

public class ReporterFactory
{
  private static HashMap<String, Class<? extends ApprovalFailureReporter>> reporters = new HashMap<String, Class<? extends ApprovalFailureReporter>>();
  public static class FileTypes
  {
    public static final String  Text    = "txt";
    public static final String  Html    = "html";
    public static final String  Excel   = "csv";
    public static final String  File    = "file";
    public static final String  Image   = "png";
    private static final String Default = "default";
  }
  static
  {
    setupReporters();
  }
  public static ApprovalFailureReporter get(String string)
  {
    ApprovalFailureReporter returned = getFromAnnotation();
    returned = tryFor(returned, reporters.get(string));
    returned = tryFor(returned, reporters.get(FileTypes.Default));
    return returned;
  }
  public static ApprovalFailureReporter getFromAnnotation()
  {
    UseReporter reporter = getAnnotationFromStackTrace(UseReporter.class);
    return reporter == null ? null : getReporter(reporter);
  }
  private static ApprovalFailureReporter getReporter(UseReporter reporter)
  {
    Class<? extends ApprovalFailureReporter>[] classes = reporter.value();
    List<ApprovalFailureReporter> reporters = new ArrayList<ApprovalFailureReporter>();
    for (Class<? extends ApprovalFailureReporter> clazz : classes)
    {
      ApprovalFailureReporter instance = ClassUtils.create(clazz);
      reporters.add(instance);
    }
    return reporters.size() == 1 ? reporters.get(0) : new MultiReporter(reporters);
  }
  private static <T extends Annotation> T getAnnotationFromStackTrace(Class<T> annotationClass)
  {
    StackTraceElement[] trace = Thread.currentThread().getStackTrace();
    for (StackTraceElement stack : trace)
    {
      Method method = null;
      Class<?> clazz = null;
      try
      {
        String methodName = stack.getMethodName();
        clazz = Class.forName(stack.getClassName());
        method = clazz.getMethod(methodName, null);
      }
      catch (Exception e)
      {
        //ignore
      }
      T annotation = null;
      if (method != null)
      {
        annotation = method.getAnnotation(annotationClass);
      }
      if (annotation != null) { return annotation; }
      annotation = clazz.getAnnotation(annotationClass);
      if (annotation != null) { return annotation; }
    }
    return null;
  }
  private static ApprovalFailureReporter tryFor(ApprovalFailureReporter returned,
      Class<? extends ApprovalFailureReporter> trying)
  {
    if (returned == null && trying != null) { return ClassUtils.create(trying); }
    return returned;
  }
  private static void setupReporters()
  {
    reporters.put(FileTypes.Text, DiffReporter.class);
    reporters.put(FileTypes.Html, DiffReporter.class);
    reporters.put(FileTypes.Excel, FileLauncherReporter.class);
    reporters.put(FileTypes.File, FileLauncherReporter.class);
    reporters.put(FileTypes.Image, ImageReporter.class);
    reporters.put(FileTypes.Default, QuietReporter.class);
  }
  public static void clearAllReportersExceptDefault()
  {
    Class all = reporters.get(FileTypes.Default);
    reporters.clear();
    reporters.put(FileTypes.Default, all);
  }
}
