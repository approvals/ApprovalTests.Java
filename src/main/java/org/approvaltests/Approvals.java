package org.approvaltests;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.hadoop.fs.Path;
import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureOverrider;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalBinaryFileWriter;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.DirectoryToDirectoryWriter;
import org.approvaltests.writers.FileApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.approvaltests.writers.ResultSetApprovalWriter;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.lambda.functions.Function1;
import org.lambda.functions.implementations.F1;
import org.lambda.query.Query;

import com.spun.util.ArrayUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.images.ImageWriter;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.SqlLoader;

public class Approvals
{
  /**
   * @deprecated Use verify()
   */
  public static void approve(String response) throws Exception
  {
    verify(response);
  }
  public static void verify(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
  }
  public static void verify(Object o) throws Exception
  {
    verify("" + o);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, T[] array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, T[] array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, Iterable<T> array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, Iterable<T> array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Component c) throws Exception
  {
    verify(c);
  }
  public static void verify(Component c) throws Exception
  {
    verify(new ComponentApprovalWriter(c), FileTypes.Image);
  }
  /**
   * @deprecated Use verifyHtml()
   */
  public static void approveHtml(String response) throws Exception
  {
    verifyHtml(response);
  }
  public static void verifyHtml(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(File generateFile)
  {
    verify(generateFile);
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile), FileTypes.File);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Image image)
  {
    verify(image);
  }
  public static void verify(Image image)
  {
    approve(ImageWriter.toBufferedImage(image), createApprovalNamer());
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(BufferedImage bufferedImage)
  {
    verify(bufferedImage);
  }
  public static void verify(BufferedImage bufferedImage)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(writter, namer, reporter);
  }
  public static void verify(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writter, namer), reporter);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, String fileType)
  {
    verify(writter, fileType);
  }
  public static void verify(ApprovalWriter writter, String fileType)
  {
    verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
  }
  /**
   * @deprecated Use verifyXml()
   */
  public static void approveXml(String xml) throws Exception
  {
    verifyXml(xml);
  }
  public static void verifyXml(String xml) throws Exception
  {
    verify(new ApprovalXmlWriter(xml), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(FileApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, reporter);
  }
  public static void verify(FileApprover approver, ApprovalFailureReporter reporter)
  {
    try
    {
      if (!approver.approve())
      {
        boolean passed = false;
        if (reporter instanceof ApprovalFailureOverrider)
        {
          passed = approver.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
        }
        if (!passed)
        {
          approver.reportFailure(reporter);
          approver.fail();
        }
        else
        {
          approver.cleanUpAfterSuccess(reporter);
        }
      }
      else
      {
        approver.cleanUpAfterSuccess(reporter);
      }
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ExecutableQuery query) throws Exception
  {
    verify(query);
  }
  public static void verify(ExecutableQuery query) throws Exception
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), new JUnitStackTraceNamer(),
        new ExecutableQueryFailure(query));
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Map map) throws Exception
  {
    verify(map);
  }
  public static void verify(Map map) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(RackResponse response) throws Exception
  {
    verify(response);
  }
  public static void verify(RackResponse response) throws Exception
  {
    if (isImage(response))
    {
      String fileType = "png";
      verify(new ApprovalBinaryFileWriter(response.getResponse(), fileType), fileType);
    }
    else
    {
      verifyHtml(response.getResponse().toString());
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ResultSet rs) throws Exception
  {
    verify(rs);
  }
  public static void verify(ResultSet rs) throws Exception
  {
    verify(new ResultSetApprovalWriter(rs), "csv");
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(SqlLoader loader) throws Exception
  {
    verify(loader);
  }
  public static void verify(SqlLoader loader) throws Exception
  {
    verify(new SqlLoader.ExecutableWrapper(loader));
  }
  private static boolean isImage(RackResponse response)
  {
    String type = response.getHeaders().get(RackResponseUtils.CONTENT_TYPE);
    return RackResponseUtils.CONTENT_TYPE_IMAGE.equals(type);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return new JUnitStackTraceNamer();
  }
  private static void approve(BufferedImage bufferedImage, ApprovalNamer namer)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + Path.SEPARATOR + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory), FileTypes.File);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, new F1<File, String>(mismatched.get(1))
    {
      {
        ret(a.getName());
      }
    }).toString();
  }
}