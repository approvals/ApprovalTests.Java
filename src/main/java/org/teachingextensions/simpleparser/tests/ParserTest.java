package org.teachingextensions.simpleparser.tests;

import junit.framework.TestCase;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.teachingextensions.simpleparser.Parser;

@UseReporter(DiffReporter.class)
public class ParserTest extends TestCase
{
  private int    a = 1;
  private String b = "howdy";
  public void testSimpleParse() throws Exception
  {
    String text = Parser.parse("This should show \r\n  a = {a} \r\n  b = {b}", this);
    Approvals.verify(text);
  }
}
