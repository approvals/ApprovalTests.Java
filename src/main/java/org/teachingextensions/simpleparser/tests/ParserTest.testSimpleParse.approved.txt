This should show 
  a = 1 
  b = howdy